### PR TITLE
Miscellaneous fixes

### PR DESCRIPTION
**Pull request by DemiMarie**
This is miscellaneous fixes that were easy to implement.

### CODE REVIEW

**Comment by marmarek** [comment]
This appears to break several things. Unfortunately automatic test result collection didn't work (unrelated openQA issue), so here is the full log: https://openqa.qubes-os.org/tests/23199/file/system_tests-tests-qubes.tests.extra.log

Example failures:
```
=====================================================================
FAIL: splitgpg.tests/TC_00_Direct_debian-11/test_080_option_parser
Check if split-gpg agrees with gpg about options parsing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 367, in test_080_option_parser
    splitgpg_needs_arg = self._check_if_options_takes_argument(
  File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 348, in _check_if_options_takes_argument
    self.fail(
AssertionError: qubes-gpg-client --list-key have not complained about --garbage options: Abbreviated option ‘--list-key’ must be written as ‘--list-keys’


======================================================================
FAIL: splitgpg.tests/TC_00_Direct_fedora-34/test_031_sign_verify_detached
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 161, in test_031_sign_verify_detached
    self.assertIn('\ngpg: Good signature from', verification_result.decode())
AssertionError: '\ngpg: Good signature from' not found in "IDs must not start with '-'\nEOF\n"

======================================================================
FAIL: splitgpg.tests/TC_00_Direct_fedora-34/test_050_sign_verify_files
Test for --output option
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 269, in test_050_sign_verify_files
    self.assertEquals(decoded_msg.decode(), msg)
AssertionError: '' != 'Test message'
+ Test message
```

The first one is because `--list-key` is a valid alias for the `--list-keys` option, not only abbreviation.

**Comment by DemiMarie** [comment]
> This appears to break several things. Unfortunately automatic test result collection didn't work (unrelated openQA issue), so here is the full log: https://openqa.qubes-os.org/tests/23199/file/system_tests-tests-qubes.tests.extra.log
> 
> Example failures:
> 
> ```
> =====================================================================
> FAIL: splitgpg.tests/TC_00_Direct_debian-11/test_080_option_parser
> Check if split-gpg agrees with gpg about options parsing
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 367, in test_080_option_parser
>     splitgpg_needs_arg = self._check_if_options_takes_argument(
>   File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 348, in _check_if_options_takes_argument
>     self.fail(
> AssertionError: qubes-gpg-client --list-key have not complained about --garbage options: Abbreviated option ‘--list-key’ must be written as ‘--list-keys’
> 
> 
> ======================================================================
> FAIL: splitgpg.tests/TC_00_Direct_fedora-34/test_031_sign_verify_detached
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 161, in test_031_sign_verify_detached
>     self.assertIn('\ngpg: Good signature from', verification_result.decode())
> AssertionError: '\ngpg: Good signature from' not found in "IDs must not start with '-'\nEOF\n"
> 
> ======================================================================
> FAIL: splitgpg.tests/TC_00_Direct_fedora-34/test_050_sign_verify_files
> Test for --output option
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/site-packages/splitgpg/tests.py", line 269, in test_050_sign_verify_files
>     self.assertEquals(decoded_msg.decode(), msg)
> AssertionError: '' != 'Test message'
> + Test message
> ```
> 
> The first one is because `--list-key` is a valid alias for the `--list-keys` option, not only abbreviation.

I will fix the first one, and will change the other code to not complain about a leading `-` if it is the only character in an option.  I will also add UTF-8 validity checks.

File: src/gpg-common.c [hunk 1]
```diff
@@ -115,34 +137,194 @@ void handle_opt_verify(char *untrusted_sig_path, int *list, int *list_count, int
...
+            else
+                assert(optind - lastind == 1);
+            if (delta > len || memcmp(optname, optval, delta)) {
+                fprintf(stderr,
+                        "split-gpg: internal error: option misparsed by getopt_long(3)\n");
+                abort();
+            }
+            if (delta < len) {
+                fprintf(stderr,
+                        "Abbreviated option ‘--%.*s’ must be written as ‘--%s’\n",
```
Comment (marmarek) [inline src/gpg-common.c:321]: Please avoid using non-ASCII characters in the error messages.

File: src/gpg-common.c [hunk 2]
```diff
@@ -195,6 +377,19 @@ int parse_options(int argc, char *untrusted_argv[], int *input_fds,
         }
 
     }
+    // Only allow key IDs to begin with ‘-’ if the options list was terminated by ‘--’,
+    // or if the argument is a literal “-” (which is never considered an option)
+    if (!lastarg || strcmp(lastarg, "--")) {
+        for (int i = optind; i < argc; ++i) {
+            const char *const untrusted_arg = untrusted_argv[i];
+            if (untrusted_arg[0] == '-' && untrusted_arg[1]) {
+                fprintf(stderr, u8"Non-option arguments must not start with ‘-’, unless preceeded by “--”\n"
```
Comment (marmarek) [inline src/gpg-common.c:386]: Here too.

**Review by marmarek — CHANGES_REQUESTED** [review]
Besides the below, tests still fail with `qubes-gpg-client --list-sig have not complained about --garbage options: Abbreviated option ‘--list-sig’ must be written as ‘--list-sigs’`

**Comment by marmarek** [comment]
> The way I debugged this in the past was by changing the gpg-wrapper to echo all commands to a file,

I did the same here. And got this:
```
qubes-gpg-client-wrapper --enable-special-filenames --batch --no-sk-comments --status-fd 99 --no-tty --charset utf8 --enable-progress-filter --exit-on-status-write-error --display :0 --logger-fd 101 --sign --detach --armor --textmode -u 465649BEF6CF08B0 -- '-&102'
Abbreviated option '--detach' must be written as '--detach-sign'
```

**Comment by DemiMarie** [comment]
PipelineRetry

**Review by marmarek — APPROVED** [review]

**Pull request merged**